Fix for app:enable in case of not valid/not existing enterprise key

pako81 opened the pull request:
## Description

This PR fixes the current behaviour of the `app:enable` occ command in case of not valid/not existing enterprise key.

## Related Issue

- Fixes https://github.com/owncloud/core/issues/36351

## Motivation and Context

Currently, when running the `app:enable` occ command for enabling an enterprise app and the enterprise key is not valid/not existing, the command does not return any warning/error message. Still, the enterprise app stays disabled. Only when looking into the owncloud.log file one will see an error message about the invalid license key.

## How Has This Been Tested?

Manually, by: - disabling an enterprise app, - setting an invalid enteprise key, - trying to re-enable that enterprise app. Now, the `$appId cannot be enabled because of invalid enteprise key` message is shown.

Implemented for both cases in case groups are/are not given as input to the command.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised 
- [x] Changelog item

Commit:
Fix for app:enable in case of not valid/not existing enterprise key

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=36399) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=36399) before we can accept your contribution.<br/><hr/>**pastripodi@owncloud.com** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/core?pullRequest=36399) it.</sub>

phil-davis commented:
https://drone.owncloud.com/owncloud/core/21265/3/6
```
enabling apps
+ echo 'enabling apps'
+ php occ app:enable files_sharing
An unhandled exception has been thrown:
Error: Class 'OCA\Enterprise_Key\EnterpriseKey' not found in /drone/src/core/Command/App/Enable.php:74
```
It still needs to work for the community edition, when there is no `enterprise_key` app.

micbar commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +71,19 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 
 		$groups = $input->getOption('groups');
+		$key = new \OCA\Enterprise_Key\EnterpriseKey(false, \OC::$server->getConfig());
 		if (empty($groups)) {
```
Please inject config in constructor.

pako81 replied:
will do

micbar commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +77,21 @@ protected function execute(InputInterface $input, OutputInterface $output) {
-		if (empty($groups)) {
+		if (\OC_App::isEnabled('enterprise_key')) {
+			$key = new \OCA\Enterprise_Key\EnterpriseKey(false, \OC::$server->getConfig());
+		}	
```
use the global property here: $this->config

pako81 commented:
better now?

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +77,21 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		}
 		if (empty($groups)) {
+			if ($key && !$key->isValid()) {
+				$output->writeln($appId . ' cannot be enabled because of invalid enteprise key');
```
```suggestion
				$output->writeln($appId . ' cannot be enabled because of invalid enterprise key');
```

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +77,21 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 			$output->writeln($appId . ' enabled');
 		} else {
+			if ($key && !$key->isValid()) {
+				$output->writeln($appId . ' cannot be enabled because of invalid enteprise key');
```
```suggestion
				$output->writeln($appId . ' cannot be enabled because of invalid enterprise key');
```

karakayasemi commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +77,21 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 
 		$groups = $input->getOption('groups');
+		if (\OC_App::isEnabled('enterprise_key')) {
+			$key = new \OCA\Enterprise_Key\EnterpriseKey(false, $this->config);
```
To prevent creating duplicate code block, you can validate the key directly after `$key` assignment. Also, in this way, you do not need to check whether the key exists.

pako81 replied:
The duplicate code block is actually needed depending if the `$groups` option has been given or not as input to the occ command, and it was already there before this PR. Or did I misunderstand the suggestion?

karakayasemi replied:
As the last step, just run `make test-php-style-fix` for code style adjustment.

micbar replied:
The enterprise key app is not enabled in CI. Therefore it fails.
@phil-davis

phil-davis replied:
Looking at it - when can likely give the code analysis tool(s) a clue to skip that line

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -71,10 +77,21 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+			$key = new \OCA\Enterprise_Key\EnterpriseKey(false, $this->config);
+		}
 		if (empty($groups)) {
+			if ($key && !$key->isValid()) {
```
He means that this `if` test can be moved to before `if (empty($groups))` and then it is only needed once.

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -70,6 +76,14 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		}
 
+		if (\OC_App::isEnabled('enterprise_key')) {
+		$key = new \OCA\Enterprise_Key\EnterpriseKey(false, $this->config);
```
Indent is strange

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -70,6 +76,14 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		$key = new \OCA\Enterprise_Key\EnterpriseKey(false, $this->config);
+		        if ($key && !$key->isValid()) {
+			$output->writeln($appId . ' cannot be enabled because of invalid enterprise key');
+			return 1;
```
And these lines need indent inside the `if` block

phil-davis commented:
@micbar is this progressing?
(could have an acceptance test added if it is planned for release...)

phil-davis commented:
@pako81 @micbar I  rebased and fixed the code-style fail.
Let's see if the "real" CI passes...

phil-davis committed:
Ignore EnterpriseKey class not found by code analysis tools

suppress PhanUndeclaredClassMethod for \OCA\Enterprise_Key\EnterpriseKey
Ignore OCA\Enterprise_Key\EnterpriseKey not found for phpstan

phil-davis committed:
Adjust AppsEnableTest for Enable constructor params

phil-davis commented:
And adjusted the unit test so it knows about the new param to the constructor.
Maybe this time CI will pass?

micbar commented:
> And adjusted the unit test so it knows about the new param to the constructor.
> Maybe this time CI will pass?

PHpstan fails because the code of Enterprise Key is not available.

phil-davis committed:
Changelog for 36399 app enable with invalid enterprise key

phil-davis commented:
> > And adjusted the unit test so it knows about the new param to the constructor.
> > Maybe this time CI will pass?
> 
> PHpstan fails because the code of Enterprise Key is not available.

I "fixed" that by making it ignore that - phpstan CI is passing.

phil-davis commented:
I added a changelog. I think this is ready for review.

There are existing `cliProvisioning/enableApp.feature` tests that cover enabling an ordinary app - so that ensures that this code change did not break anything for ordinary apps.

We are not easily able to get ourselves an enterprise app and the enterprise_key app etc during core CI. So this change will need manual testing during the release QA process.

phil-davis commented on core/Command/App/Enable.php:
```diff
@@ -70,6 +76,16 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 			return 1;
 		}
 
+		if (\OC_App::isEnabled('enterprise_key')) {
```
The issue we found now is that if you have the enterprise app enabled, but have an invalid enterprise key, and you are just trying to enable a  non-enterprise app, then the app cannot be enabled. e.g. enabling the `testing` app does not work.

If there is some way to check if the app being enabled is an Enterprise app, then this could be fixed by:
```
if (\OC_App::isEnterpriseApp($appId) && \OC_App::isEnabled('enterprise_key')) {
```
